message/mod.ts

```diff
@@ -0,0 +1,110 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+// This module is browser compatible.
+/**
+ * Messaging interface designed to scale to hundreds of modules.
+ * @module
+ */
+import { sprintf } from "../fmt/printf.ts";
+
+/**
+ * Message logging levels.
+ *
+ * Different message levels:
+ *
+ * + `log` - Debug messages, mostly used for developers.
+ * + `info` - Module status information, suitable for users.
+ * + `warn` - Warning messages about something that can't be ignored.
+ * + `error` - Something has gone wrong and the module cannot continue.
+ */
+export type MessageLevel = "log" | "info" | "warn" | "error";
+
+/** Metadata about a message. */
+export interface MessageContext {
+  /** Category sending this message. */
+  category: MessageCategory;
+  /** Message level/severity. */
+  level: MessageLevel;
+  /** Filename (typically `import.meta.url`) of the message sender. */
+  filename: string;
+  /** Time the message was sent. */
+  time: Date;
+}
+
+/** A class for categorized messages. */
+export class MessageCategory {
+  /** Silence all output from this category if true. */
+  quiet = false;
+
+  constructor(public readonly name: string) {}
+
+  /** Log a debug message. */
+  log(arg: string, ...args: unknown[]) {
```

Comment:
In deno, `log()` and `info()` log messages in the same level (1), and `debug()` logs messages in lower level (0). ref. https://github.com/denoland/deno/blob/f729576b2db2aa6ce000a598ad2e45533f686213/ext/console/02_console.js#L1977-L2005

I think these methods should be align with it.